test.ml

```diff
@@ -11,117 +11,145 @@
-let Ok (content_txt, _) = to_result decoder
-let decoder = decoding decoder (Message.Decoder.p_discard_part content)
-let Ok (`Next : [`Next | `End ]) = to_result decoder
-let decoder = decoding decoder (Message.Decoder.p_next_part content)
-let Ok (content_img, _) = to_result decoder
-
-let bound       = Message.Decoder.p_bound_of_content content
-let decoder_b64 = Base64.Convenience.decoder bound (decoder_src decoder)
-
-let rec b64_to_result decoder_b64 =
-  let open Base64.Convenience in
-  match decode decoder_b64 with
+    (* Read another chunk, *)
+    let n = read_input () in
+    (* hand it to the decoder, *)
+    Convenience.src decoder tmp 0 n;
+    (* and carry on parsing *)
+    get decoder
+  | `Done v -> v
+  | `Error exn -> raise (MrMimeError exn)
+
+(* [get_value] runs the [decoder] and raises [Failure] *)
+let get_value v err (decoder : ('input, 'a) Convenience.decoder) =
+  if get decoder <> v then
+    failwith err
+
+(* Avoid having the GIF file printed on the console if you run this interactively *)
+let print_string =
+  if Unix.isatty Unix.stdout then
+    ignore
+  else
+    print_string
+
+(* The Base64 module includes functions as in the general Convenience module, but with a more
+   detailed return type, to allow for best-effort decoding of corrupted Base64 files. This function
+   is therefore similar in structure to [get]. *)
+let rec get_b64 decoder_b64 =
+  match Base64.decode decoder_b64 with
   | `Continue ->
-    let n = read_into ~newline:LF ch tmp 0 1024 in
-    src decoder_b64 tmp 0 n;
-    b64_to_result decoder_b64
+    (* Read another chunk, *)
+    let n = read_input () in
+    (* hand it to the Base64 decoder, *)
+    Base64.src decoder_b64 tmp 0 n;
+    (* and carry on decoding it *)
+    get_b64 decoder_b64
   | `String s ->
+    (* A chunk of decoded data *)
     print_string s;
-    b64_to_result decoder_b64
+    get_b64 decoder_b64
   | `End s ->
-    print_string s;
-    Ok ()
+    (* Decoding complete *)
+    print_string s
   | `Dirty s ->
-    b64_to_result decoder_b64
-  | `Error exn -> Error exn
-
-let Ok () = b64_to_result decoder_b64
-
-let decoder = decoding decoder (Message.Decoder.p_end_of_part content)
-let Ok (`End : [ `Next | `End ]) = to_result decoder
+    (* Ignore this chunk and carry on (shouldn't happen in this demo!) *)
+    get_b64 decoder_b64
+  | `Error exn -> raise (MrMimeError exn)
+
+(* Now let's process the actual input message. The most general way of doing this is to use a
+   Message.Decoder.p_message which returns a structure containing the entire message fully parsed:
+
+   let message = get (Convenience.decoder (Input.create_bytes 4096) Message.Decoder.p_message)
+
+   However, it's also possible to walk the structure of the message, only decoding the parts
+   required. For this, we start off with a Message.Decoder.p_header:
+*)
+
+let decoder = Convenience.decoder (Input.create_bytes 4096) Message.Decoder.p_header
+(* XXX What are the fields? *)
```

Comment:
Here, and in various further places, I couldn't figure out exactly what the list is (the actual value is `[]` in all these case). What is the list?